### ISSUE TITLE
Kubernetes namespaces stuck in terminating state

### ISSUE THREAD

**Issue by paralin** [kubernetes/kubernetes#19317]
I tried to delete some namespaces from my kubernetes cluster, but they've been stuck in Terminating state for over a month.

```
kubectl get ns
NAME              LABELS    STATUS        AGE
myproject         <none>    Active        12d
default              <none>    Active        40d
anotherproject  <none>    Terminating   40d
openshift         <none>    Terminating   40d
openshift-infra   <none>    Terminating   40d
```

The openshift namespaces were made as part of the example in this repo for running Openshift under Kube. 

There's nothing in any of these namespaces (I used get on every resource type and they're all empty).

So what's holding up the terminate?

The kube cluster is healthy:

```
NAME                 STATUS    MESSAGE              ERROR
etcd-1               Healthy   {"health": "true"}   
controller-manager   Healthy   ok                   
scheduler            Healthy   ok                   
etcd-0               Healthy   {"health": "true"}   
```

The versions are:

```
Client Version: version.Info{Major:"1", Minor:"2+", GitVersion:"v1.2.0-alpha.4.208+c39262c9915b0b", GitCommit:"c39262c9915b0b1c493de66f37c49f3ef587cd97", GitTreeState:"clean"}
Server Version: version.Info{Major:"1", Minor:"2+", GitVersion:"v1.2.0-alpha.4.166+d9ab692edc08a2", GitCommit:"d9ab692edc08a279396b29efb4d7b1e6248dfb60", GitTreeState:"clean"}
```

The server version corresponds to this commit: https://github.com/paralin/kubernetes/commit/d9ab692edc08a279396b29efb4d7b1e6248dfb60

Compiled from source. Cluster was built using kube-up to GCE with the following env:

```
export KUBERNETES_PROVIDER=gce
export KUBE_GCE_ZONE=us-central1-b
export MASTER_SIZE=n1-standard-1
export MINION_SIZE=n1-standard-2
export NUM_MINIONS=3

export KUBE_ENABLE_NODE_AUTOSCALER=true
export KUBE_AUTOSCALER_MIN_NODES=3
export KUBE_AUTOSCALER_MAX_NODES=3

export KUBE_ENABLE_DAEMONSETS=true
export KUBE_ENABLE_DEPLOYMENTS=true

export KUBE_ENABLE_INSECURE_REGISTRY=true
```

Any ideas?

**Comment by ncdc** [comment]
cc @derekwaynecarr. Do you think the namespace controller is in some sort of infinite loop?

**Comment by derekwaynecarr** [comment]
Can you paste the output for:

kubectl get namespace/openshift -o json

I assume openshift is no longer running on your cluster?  Is there any
content in that namespace?

**Comment by lavalamp** [comment]
Smells like different components have different ideas about the finalizer list? Does rebooting controller-manager change anything?

**Comment by paralin** [comment]
kubectl get ns openshift -o json

```
{
    "kind": "Namespace",
    "apiVersion": "v1",
    "metadata": {
        "name": "openshift",
        "selfLink": "/api/v1/namespaces/openshift",
        "uid": "0a659292-94af-11e5-855c-42010af00002",
        "resourceVersion": "14645862",
        "creationTimestamp": "2015-11-27T02:32:01Z",
        "deletionTimestamp": "2015-12-25T03:20:25Z",
        "annotations": {
            "openshift.io/sa.scc.mcs": "s0:c6,c0",
            "openshift.io/sa.scc.supplemental-groups": "1000030000/10000",
            "openshift.io/sa.scc.uid-range": "1000030000/10000"
        }
    },
    "spec": {
        "finalizers": [
            "openshift.io/origin"
        ]
    },
    "status": {
        "phase": "Terminating"
    }
}
```

Interestingly the finalizer is set to openshift.io/origin.

I tried deleting the finalizer out of the namespace using kubectl edit, but it still remains in another get operation.

**Comment by paralin** [comment]
This also happens with the one other namespace I manually created in OpenShift with the projects system:

```
Error from server: Namespace "dotabridge-dev" cannot be updated: The system is ensuring all content is removed from this namespace.  Upon completion, this namespace will automatically be purged by the system.
```

I'm not actually using OpenShift anymore so these namespaces are pretty much stuck in my prod cluster until I can figure out how to get past this.

**Comment by paralin** [comment]
Deleted the controller-manager pod and the associated pause pod and restarted kubelet on the master. The containers were re-created, `kubectl get cs` shows everything as healthy, but the namespaces remain.

**Comment by davidopp** [comment]
@kubernetes/rh-cluster-infra

**Comment by derekwaynecarr** [comment]
@paralin - there is no code issue, but maybe I can look to improve in the openshift example clean-up scripts or document the steps.  When you created a project in openshift, it created a namespace for that project, and annotated the namespace with a `finalizer` token that said before a namespace is deleted, an external agent needs to remove its lock on the object that says it was done clean-up.  Since you are no longer running openshift, its agent did not remove the lock and take part in the termination flow.

A quick fix:

```
# find each namespace impacted
$ kubectl get namespaces -o json | grep "openshift.io/origin"
$ kubect get namespace <ns> -o json > temp.json
# vi temp.json and remove the finalizer entry for "openshift.io/origin"
$ curl -H "Content-Type: application/json" -X PUT --data-binary @temp.json http://127.0.0.1:8080/api/v1/namespaces/<name_of_namespace>/finalize
# wait a moment, and you should see your namespace removed
$ kubectl get namespaces 
```

That will remove the lock that blocks the namespace from being completely terminated, and you should quickly see that the namespace is removed from your system.

Closing the issue, but feel free to comment if you continue to have problems or hit me up on slack.

**Comment by bkmagnetron** [comment]
I'm facing the same issue
```
# oc version
oc v1.3.0-alpha.2
kubernetes v1.3.0-alpha.1-331-g0522e63
```
I have deleted the project named "gitlab" via web console.
```
# kubectl get namespace/gitlab -o json
{
    "kind": "Namespace",
    "apiVersion": "v1",
    "metadata": {
        "name": "gitlab",
        "selfLink": "/api/v1/namespaces/gitlab",
        "uid": "cd86c372-481e-11e6-aebc-408d5c676116",
        "resourceVersion": "3115",
        "creationTimestamp": "2016-07-12T10:53:01Z",
        "deletionTimestamp": "2016-07-12T11:11:36Z",
        "annotations": {
            "openshift.io/description": "",
            "openshift.io/display-name": "GitLab",
            "openshift.io/requester": "developer",
            "openshift.io/sa.scc.mcs": "s0:c8,c7",
            "openshift.io/sa.scc.supplemental-groups": "1000070000/10000",
            "openshift.io/sa.scc.uid-range": "1000070000/10000"
        }
    },
    "spec": {
        "finalizers": [
            "kubernetes"
        ]
    },
    "status": {
        "phase": "Terminating"
    }
}
```

As said by @derekwaynecarr I did the following
```
# kubectl get namespace gitlab -o json > temp.json
```
```
# cat temp.json
```
```
{
    "kind": "Namespace",
    "apiVersion": "v1",
    "metadata": {
        "name": "gitlab",
        "selfLink": "/api/v1/namespaces/gitlab",
        "uid": "cd86c372-481e-11e6-aebc-408d5c676116",
        "resourceVersion": "3115",
        "creationTimestamp": "2016-07-12T10:53:01Z",
        "deletionTimestamp": "2016-07-12T11:11:36Z",
        "annotations": {
            "openshift.io/description": "",
            "openshift.io/display-name": "GitLab",
            "openshift.io/requester": "developer",
            "openshift.io/sa.scc.mcs": "s0:c8,c7",
            "openshift.io/sa.scc.supplemental-groups": "1000070000/10000",
            "openshift.io/sa.scc.uid-range": "1000070000/10000"
        }
    },
    "spec": {
        "finalizers": [
            "kubernetes"   <---removed
        ]
    },
    "status": {
        "phase": "Terminating"
    }
}
```
and
```
# curl -k -H "Content-Type: application/json" -X PUT --data-binary @temp.json https://10.28.27.65:8443/api/v1/namespaces/gitlab/finalize
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "User \"system:anonymous\" cannot update namespaces/finalize in project \"gitlab\"",
  "reason": "Forbidden",  <--- seems like nothing happened
  "details": {
    "name": "gitlab",
    "kind": "namespaces/finalize"
  },
  "code": 403
}
```
but it is removed.

**Comment by jchauncey** [comment]
Im facing the same problem in GKE. Bouncing the cluster definitely fixes the issue (they are immediately terminated).

**Comment by linfan** [comment]
I believe this issue still exist in v1.3 release.

Manually remove the finalizer doesn't seems to help.

```
$ curl -H "Content-Type: application/json" -X PUT --data-binary @temp.json http://ip-172-31-14-177:8080/api/v1/namespaces/limit/finalize
{
  "kind": "Namespace",
  "apiVersion": "v1",
  "metadata": {
    "name": "limit",
    "selfLink": "/api/v1/namespaces/limit/finalize",
    "uid": "caf5daa5-57f8-11e6-9e7e-0ad69bcef303",
    "resourceVersion": "10171",
    "creationTimestamp": "2016-08-01T15:01:14Z",
    "deletionTimestamp": "2016-08-02T04:30:24Z"
  },
  "spec": {},
  "status": {
    "phase": "Terminating"
  }
```

Several hours, it still remain.

```
$ kubectl get namespaces
NAME                STATUS         AGE
...                 ...
limit               Terminating   13h
```

Until I completely restart the master server, all "terminating" namespaces gone...

**Comment by monaka** [comment]
Still v1.4.0 also...

```
$ kubectl get ns openmct -o json
{
    "kind": "Namespace",
    "apiVersion": "v1",
    "metadata": {
        "name": "openmct",
        "selfLink": "/api/v1/namespaces/openmct",
        "uid": "34124209-8e8d-11e6-8260-000d3a505da6",
        "resourceVersion": "11957259",
        "creationTimestamp": "2016-10-10T01:59:39Z",
        "deletionTimestamp": "2016-10-10T02:13:46Z",
        "labels": {
            "heritage": "deis"
        }
    },
    "spec": {
        "finalizers": [
            "kubernetes"
        ]
    },
    "status": {
        "phase": "Terminating"
    }
}
```

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.0", GitCommit:"a16c0a7f71a6f93c7e0f222d961f4675cd97a46b", GitTreeState:"clean", BuildDate:"2016-09-26T18:16:57Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.0+coreos.2", GitCommit:"672d0ab602ada99c100e7f18ecbbdcea181ef008", GitTreeState:"clean", BuildDate:"2016-09-30T05:49:34Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by jsloyer** [comment]
im hitting the error with 1.3.5 as well....

```
$kubectl version
Client Version: version.Info{Major:"1", Minor:"3", GitVersion:"v1.3.5", GitCommit:"b0deb2eb8f4037421077f77cb163dbb4c0a2a9f5", GitTreeState:"clean", BuildDate:"2016-08-11T20:29:08Z", GoVersion:"go1.6.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"3", GitVersion:"v1.3.6", GitCommit:"ae4550cc9c89a593bcda6678df201db1b208133b", GitTreeState:"clean", BuildDate:"2016-09-22T01:52:27Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by monaka** [comment]
At least in my case, it might be API issue...?

```
$ kubectl get ns | grep Terminating | wc -l
7
```

kube-apiserver:

```
E1017 01:47:42.450498       1 errors.go:63] apiserver received an error that is not an unversioned.Status: no kind "DeleteOptions" is registered for version "net.alpha.kubernetes.io/v1alpha1"
E1017 01:47:42.921381       1 errors.go:63] apiserver received an error that is not an unversioned.Status: no kind "DeleteOptions" is registered for version "net.alpha.kubernetes.io/v1alpha1"
E1017 01:47:47.462928       1 errors.go:63] apiserver received an error that is not an unversioned.Status: no kind "DeleteOptions" is registered for version "net.alpha.kubernetes.io/v1alpha1"
E1017 01:47:47.464947       1 errors.go:63] apiserver received an error that is not an unversioned.Status: no kind "DeleteOptions" is registered for version "net.alpha.kubernetes.io/v1alpha1"
E1017 01:47:52.713584       1 errors.go:63] apiserver received an error that is not an unversioned.Status: no kind "DeleteOptions" is registered for version "net.alpha.kubernetes.io/v1alpha1"
```

kube-controller-manager

```
E1017 01:50:36.002533       1 namespace_controller.go:163] no kind "DeleteOptions" is registered for version "net.alpha.kubernetes.io/v1alpha1"
E1017 01:50:36.040668       1 namespace_controller.go:163] no kind "DeleteOptions" is registered for version "net.alpha.kubernetes.io/v1alpha1"
E1017 01:50:37.066455       1 namespace_controller.go:163] no kind "DeleteOptions" is registered for version "net.alpha.kubernetes.io/v1alpha1"
E1017 01:50:37.102275       1 namespace_controller.go:163] no kind "DeleteOptions" is registered for version "net.alpha.kubernetes.io/v1alpha1"
E1017 01:50:38.229602       1 namespace_controller.go:163] no kind "DeleteOptions" is registered for version "net.alpha.kubernetes.io/v1alpha1"
E1017 01:50:38.602775       1 namespace_controller.go:163] no kind "DeleteOptions" is registered for version "net.alpha.kubernetes.io/v1alpha1"
E1017 01:50:39.181639       1 namespace_controller.go:163] no kind "DeleteOptions" is registered for version "net.alpha.kubernetes.io/v1alpha1"
```

**Comment by monaka** [comment]
In my case, ThirdPartyResource is kept on Etcd. Stucked namespaces was removed after deleting it like this.

```
etcdctl rm /registry/thirdpartyresources/default/network-policy.net.alpha.kubernetes.io
```

**Comment by zhouhaibing089** [comment]
The new issue is not the same as the original one, I think we need to create another new issue.

**Comment by zhouhaibing089** [comment]
created: https://github.com/kubernetes/kubernetes/issues/37278

**Comment by pidah** [comment]
we are hitting this issue atm on 1.4.6

**Comment by hectorj2f** [comment]
we are hitting this issue atm on 1.5.2

https://github.com/kubernetes/kubernetes/issues/37554
https://github.com/kubernetes/kubernetes/issues/37278

**Comment by hectorj2f** [comment]
I am using `v1.5.2` and the problem seems to be fixed. I am able to delete namespaces.

**Comment by nikinath** [comment]
This is fixed in `v1.5.2`. Please see: https://github.com/kubernetes/kubernetes/issues/37278#issuecomment-289780021

**Comment by paultiplady** [comment]
I'm hitting this issue on v1.6.10-gke.1. Looks like a regression. Manually deleting the finalizer fixed the problem. Here's a dump of my NS data;


```
$ kubectl get ns review-bugfix-cb-response-logging-qwil-2210 -o yaml                                             
apiVersion: v1                                                                 
kind: Namespace                                                                
metadata:                                                                      
  creationTimestamp: 2017-10-17T19:45:14Z                                      
  deletionTimestamp: 2017-10-18T22:56:53Z                                      
  name: review-bugfix-cb-response-logging-qwil-2210                            
  resourceVersion: "53696648"                                                  
  selfLink: /api/v1/namespacesreview-bugfix-cb-response-logging-qwil-2210      
  uid: b1fd43a1-b373-11e7-96b9-42010a80000b                                    
spec:                                                                          
  finalizers:                                                                  
  - kubernetes                                                                 
status:                                                                        
  phase: Terminating                                                           

$ kubectl describe ns review-bugfix-cb-response-logging-qwil-2210                                                                                                                                                                                                               
Name:           review-bugfix-cb-response-logging-qwil-2210                    
Labels:         <none>                                                         
Annotations:    <none>                                                         
Status:         Terminating                                                    

No resource quota.                                                             

No resource limits.                                   
```

**Comment by krmayankk** [comment]
i am seeing the same on 1.7.4 as well. Is the root cause the presence of tpr resources ?

**Comment by paultiplady** [comment]
I did have a TPR in all of my stuck namespaces.

**Comment by krmayankk** [comment]
@nikhita what are the symptoms to look for to make sure the terminating issue is indeed due to tpr resources ? I am seeing this issue in 1.7.4 and want to eliminate that option . Any log line to look for ?

**Comment by nikhita** [comment]
@krmayankk Sorry for the late reply. You can find some info here on _how_ it is fixed: https://github.com/kubernetes/kubernetes/issues/37554#issuecomment-289778842.

**Comment by nikhita** [comment]
Related https://github.com/kubernetes/kubernetes/issues/55002, https://github.com/kubernetes/kubernetes/issues/37554.

**Comment by attila123** [comment]
Hi, I got the same problem with Kubernetes 1.9.0 running by Minikube v0.24.1.
I started too many pods my laptop kept swapping and not responding. So I powered that off.
Upon restart lots of pods stucked, but I could delete them with `kubectl -n some_namespace delete pod --all --grace-period=0 --force` (first I deleted all the deployments, services, etc.)
I also installed rook (0.7.0) and its namespace get stuck, even after minikube restart (stop + start).
```
[vagrant@localhost ~]$ kubectl get ns rook -o json
{
    "apiVersion": "v1",
    "kind": "Namespace",
    "metadata": {
        "annotations": {
            "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"v1\",\"kind\":\"Namespace\",\"metadata\":{\"annotations\":{},\"name\":\"rook\",\"namespace\":\"\"}}\n"
        },
        "creationTimestamp": "2018-03-09T16:13:08Z",
        "deletionTimestamp": "2018-03-10T07:54:03Z",
        "name": "rook",
        "resourceVersion": "29736",
        "selfLink": "/api/v1/namespaces/rook",
        "uid": "c1a4f8e6-23b4-11e8-8129-525400ad3b43"
    },
    "spec": {
        "finalizers": [
            "kubernetes"
        ]
    },
    "status": {
        "phase": "Terminating"
    }
}
```
```
[vagrant@localhost ~]$ kubectl delete ns rook --grace-period=0 --force
warning: Immediate deletion does not wait for confirmation that the running resource has been terminated. The resource may continue to run on the cluster indefinitely.
Error from server (Conflict): Operation cannot be fulfilled on namespaces "rook": The system is ensuring all content is removed from this namespace.  Upon completion, this namespace will automatically be purged by the system.
```
I needed to move on with my work, so I stopped minikube, stopped the host VM, created a full clone of it in Virtualbox, so if there is anything I can check, any log, etc., I can provide it.
Then I `minikube delete`-ed my cluster. If I remember correctly, no resource was present in the rook namespace.

**Comment by davidmaitland** [comment]
@attila123 I was having the same issue. In the case of rook it creates a finalizer which was causing it to get stuck.

Directly from the [cleanup docs](https://rook.io/docs/rook/master/teardown.html): 

The operator is responsible for removing the finalizer after the mounts have been cleaned up. If for some reason the operator is not able to remove the finalizer (ie. the operator is not running anymore), you can delete the finalizer manually.

```
kubectl -n rook edit cluster rook
```

Look for the finalizers element and delete the following line:

```
  - cluster.rook.io
```

Now save the changes and exit the editor. Within a few seconds you should see that the cluster CRD has been deleted and will no longer block other cleanup such as deleting the rook namespace.

**Comment by iftachsc** [comment]
following @derekwaynecarr workaround - which helped me allot - wrote this script which will delete all terminating projects. sort of cleanup script. please someone tell me, i am one RedHat Openshift Enterprise! 3.7 and facing this issue. its so stupid that even i will just create an project and immediatley will delete it i will stumble on this issue. isnt this crazy?!?!!? the enterprise version cost fortune.
please just tell me i am not crazy.

this is the script; (uses jq, tested with redhat openshift ocp 3.7)
---------------------------
kubectl proxy &
serverPID=$!
for row in $(oc get ns -o json | jq -r '.items[] | select(.status.phase=="Terminating") | .metadata.name'); do
    echo "force deleting name space ${row}"
    oc project $row
    oc delete --all all,secret,pvc > /dev/null
    oc get ns $row -o json > tempns
    sed -i '' '/"kubernetes"/d' ./tempns
    curl --silent --output /dev/null -H "Content-Type: application/json" -X PUT --data-binary @tempns http://127.0.0.1:8001/api/v1/namespaces/$row/finalize
done
kill -9 $serverPID

**Comment by yevgeniyo** [comment]
Hi, I got this:


curl -H "Content-Type: application/json" -X PUT --data-binary @temp.json http://127.0.0.1:8001/api/v1/namespaces/cattle-system/finalize/
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {

  },
  "status": "Failure",
  "message": "Operation cannot be fulfilled on namespaces \"cattle-system\": the object has been modified; please apply your changes to the latest version and try again",
  "reason": "Conflict",
  "details": {
    "name": "cattle-system",
    "kind": "namespaces"
  },
  "code": 409

**Comment by akram** [comment]
I faced this issue while running oadm diagnostics NetworkChecks

the diagnostic created many projects and kept the even after successful run.
Deletion was not possible, and projects were remaining in Terminating state

The API call with curl worked also

**Comment by medanasslim** [comment]
I have already the same problem. I restarted the virtual machine, and also run "kubectl delete ns ***** --force --grace-period=0".
The namespace concerned is still stucking in Terminating.

**Comment by yan234280533** [comment]
i found that ,my event etcd is crash. So i restart my event etcd ,the namespaces are deleted soon

**Comment by nandeeshb09** [comment]
I have also same problem ,even after trying all the above scenarios.

Try to delete finalizer ,but not able too delete.

Please give the soltuion

**Comment by bhargavb7** [comment]
kubectl get namespace annoying-namespace-to-delete -o json > tmp.json
then edit tmp.json and remove"kubernetes"

curl -k -H "Content-Type: application/json" -X PUT --data-binary @tmp.json https://kubernetes-cluster-ip/api/v1/namespaces/annoying-namespace-to-delete/finalize

and it should delete your namespace,

**Comment by lyn716** [comment]
> @paralin - there is no code issue, but maybe I can look to improve in the openshift example clean-up scripts or document the steps. When you created a project in openshift, it created a namespace for that project, and annotated the namespace with a `finalizer` token that said before a namespace is deleted, an external agent needs to remove its lock on the object that says it was done clean-up. Since you are no longer running openshift, its agent did not remove the lock and take part in the termination flow.
> 
> A quick fix:
> 
> ```
> # find each namespace impacted
> $ kubectl get namespaces -o json | grep "openshift.io/origin"
> $ kubect get namespace <ns> -o json > temp.json
> # vi temp.json and remove the finalizer entry for "openshift.io/origin"
> # for example
> {
>     "kind": "Namespace",
>     "apiVersion": "v1",
>     "metadata": {
>         "name": "testing",
>         "selfLink": "/api/v1/namespaces/testing",
>         "uid": "33074e57-cb72-11e5-9d3d-28d2444e470d",
>         "resourceVersion": "234",
>         "creationTimestamp": "2016-02-04T19:05:04Z",
>         "deletionTimestamp": "2016-02-04T19:05:54Z"
>     },
>     "spec": {
>         "finalizers": [
>             "openshift.io/org"  <--- remove me
>         ]
>     },
>     "status": {
>         "phase": "Terminating"
>     }
> }
> 
> $ curl -H "Content-Type: application/json" -X PUT --data-binary @temp.json http://127.0.0.1:8080/api/v1/namespaces/<name_of_namespace>/finalize
> # wait a moment, and you should see your namespace removed
> $ kubectl get namespaces 
> ```
> That will remove the lock that blocks the namespace from being completely terminated, and you should quickly see that the namespace is removed from your system.
> 
> Closing the issue, but feel free to comment if you continue to have problems or hit me up on slack.

I think it is equal to exec `kubectl edit namespace <ns>` if uses `temp.json`.

**Comment by srikanth205** [comment]
Hi, Can any body help me out in deleting namespace which is in terminating status
PS C:\WINDOWS\system32> kubectl get ns
NAME                                           STATUS                           AGE
atlas                                               Active                           10d
config-server                                   Active                           20d
default                                             Active                              53d
dev-product-absorption                 Active                             18d
fluentd-logging                               Active                            30d
kube-public                                     Active                               53d
kube-system                                      Active                        53d
latest                                                 Active                        19d
mm-arch-iac                                   Active                            35d
monitoring                                        Terminating               3h
showcase                                          Active                           18d
test                                                Active                                  19d

**Comment by bhargavb7** [comment]
> kubectl get namespace annoying-namespace-to-delete -o json > tmp.json
> then edit tmp.json and remove"kubernetes"
> 
> curl -k -H "Content-Type: application/json" -X PUT --data-binary @tmp.json https://kubernetes-cluster-ip/api/v1/namespaces/annoying-namespace-to-delete/finalize
> 
> and it should delete your namespace,

@srikanth205 did you try using this??

**Comment by rich-nahra** [comment]
Removing finalizer worked for me.  I PUT empty spec property and removed status property completely.  Used kubectl proxy and `http://localhost:8001/api/v1/namespaces/{namespace_name}/finalizer` as the endpoint
```
{
    "kind": "Namespace",
    "apiVersion": "v1",
    "metadata": {
        "name": "testing",
        "selfLink": "/api/v1/namespaces/testing",
        "uid": "33074e57-cb72-11e5-9d3d-28d2444e470d",
        "resourceVersion": "234",
        "creationTimestamp": "2016-02-04T19:05:04Z",
        "deletionTimestamp": "2016-02-04T19:05:54Z"
    },
    "spec": {
    }
}
```

**Comment by confiq** [comment]
> curl -k -H "Content-Type: application/json" -X PUT --data-binary @tmp.json https://kubernetes-cluster-ip/api/v1/namespaces/annoying-namespace-to-delete/finalize

This didn't work on me in EKS.

**Comment by cmcarey** [comment]
Getting this issue in `1.14.3`.

I deleted a namespace, and it shows as permanently "Terminating".  Removing the finalizer made no difference, and there are no pods running in the ns.

**Comment by Bregor** [comment]
Same here in `1.15.0`

**Comment by thusihaveheard** [comment]
> Getting this issue in `1.14.3`.
> 
> I deleted a namespace, and it shows as permanently "Terminating". Removing the finalizer made no difference, and there are no pods running in the ns.

Have you solved it?

**Comment by thusihaveheard** [comment]
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.0", GitCommit:"641856db18352033a0d96dbc99153fa3b27298e5", GitTreeState:"clean", BuildDate:"2019-03-25T15:53:57Z", GoVersion:"go1.12.1", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.1-k3s.4", GitCommit:"52f3b42401c93c36467f1fd6d294a3aba26c7def", GitTreeState:"clean", BuildDate:"2019-04-15T22:13+00:00Z", GoVersion:"go1.12.1", Compiler:"gc", Platform:"linux/amd64"}

# curl -k -H “Content-Type: application/json” -X PUT –data-binary @tmp.json https://x.x.x.x:6443/api/v1/namespaces/cert-manager/finalize
curl: (6) Could not resolve host: application; Unknown error
curl: (6) Could not resolve host: xn--data-binary-lg3f; Unknown error
curl: (6) Could not resolve host: tmp.json; Unknown error
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "Unauthorized",
  "reason": "Unauthorized",
  "code": 401
}

**Comment by cmcarey** [comment]
> > Getting this issue in `1.14.3`.
> > I deleted a namespace, and it shows as permanently "Terminating". Removing the finalizer made no difference, and there are no pods running in the ns.
> 
> Have you solved it?

I have not, no.  Incredible that such an issue has been unfixed for over three years.

**Comment by rabun788** [comment]
Can we reopen this issue?

**Comment by Cobra1978** [comment]
Same Issue Here

Rancher 2.2.4

Kubernetes 1.13.5

We have a namespace stuck in Removing state, he does not have any resource inside but htere is no way to remove it

**Comment by eduardorj** [comment]
same issue here.

**Comment by difabion** [comment]
Same issue.
```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.2", GitCommit:"66049e3b21efe110454d67df4fa62b08ea79a19b", GitTreeState:"clean", BuildDate:"2019-05-16T18:55:03Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.8", GitCommit:"a89f8c11a5f4f132503edbc4918c98518fd504e3", GitTreeState:"clean", BuildDate:"2019-04-23T04:41:47Z", GoVersion:"go1.10.8", Compiler:"gc", Platform:"linux/amd64"}
```
In my case, this was a deployment of `cert-manager` that I was working into and only got as far as creating namespace, adding label, and deploying CRDs in line with Jetstack's installation docs.  Deleting the CRDs was fine but the namespace is stuck in status `Terminating`.  `kube-controller-manager` logs have this repeating ~30s:
```
W0715 22:02:18.409286       1 garbagecollector.go:647] failed to discover some groups: map[admission.certmanager.k8s.io/v1beta1:the server is currently unable to handle the request]
E0715 22:02:27.390081       1 resource_quota_controller.go:430] unable to retrieve the complete list of server APIs: admission.certmanager.k8s.io/v1beta1: the server is currently unable to handle the request
E0715 22:02:27.830397       1 memcache.go:134] couldn't get resource list for admission.certmanager.k8s.io/v1beta1: the server is currently unable to handle the request
```
No subordinate resources, ns config has a finalizer `kubernetes` in the spec and nothing else.

I'd rather it be fine with not validating this API before deleting the ns.

**Comment by difabion** [comment]
Resolved using the script from [this comment](https://github.com/kubernetes/kubernetes/issues/60807#issuecomment-497542245).

**Comment by dansl19** [comment]
in my case, it was cert-managaer ns, which stuck in termination state. It could not be delete because of CRDs which was created before but was not seen by cmd ` kubectl get all` but link to the script which has been provided by @difabion do the job. thx @difabion

**Comment by farvour** [comment]
I can't believe this issue still persists and it's a dice roll on what the actual cause is. Perhaps Kubernetes should be a little more specific showing what finalizers it is waiting on? CRDs and other namespaces should NOT cause a namespace deletion to stick. I'm flabbergasted that this is a problem since 1.8.

**Comment by singhania** [comment]
Facing the same issue on Azure kubernetes 1.13.7.

**Comment by wweir** [comment]
Same issue on aws with kops kubernetes v1.12.8

**Comment by vvoronov2** [comment]
same issue on azure k8s v1.14.1

**Comment by cmcarey** [comment]
This is clearly still an issue.  Can it be re-opened?

**Comment by marcelloromani** [comment]
The (old but effective) comment from @derekwaynecarr did the trick for me

https://github.com/kubernetes/kubernetes/issues/19317#issuecomment-180003984

the only missing step for me was `kubectl proxy` and changing the port number accordingly (8001 instead of 8080)

**Comment by iahmad-khan** [comment]
same issue on 1.15 , nothing worked in my case.

```
kubectl delete ns test
Error from server (Conflict): Operation cannot be fulfilled on namespaces "test": The system is ensuring all content is removed from this namespace.  Upon completion, this namespace will automatically be purged by the system.
```

**Comment by Slutzky** [comment]
> same issue on 1.15 , nothing worked in my case.
> 
> ```
> kubectl delete ns test

> Error from server (Conflict): Operation cannot be fulfilled on namespaces "test": The system is ensuring all content is removed from this namespace.  Upon completion, this namespace will automatically be purged by the system.
> ```



did you try to delete it using the following script ? 

```
#!/bin/bash

###############################################################################
# Copyright (c) 2018 Red Hat Inc
#
# See the NOTICE file(s) distributed with this work for additional
# information regarding copyright ownership.
#
# This program and the accompanying materials are made available under the
# terms of the Eclipse Public License 2.0 which is available at
# http://www.eclipse.org/legal/epl-2.0
#
# SPDX-License-Identifier: EPL-2.0
###############################################################################

set -eo pipefail

die() { echo "$*" 1>&2 ; exit 1; }

need() {
	which "$1" &>/dev/null || die "Binary '$1' is missing but required"
}

# checking pre-reqs

need "jq"
need "curl"
need "kubectl"

PROJECT="$1"
shift

test -n "$PROJECT" || die "Missing arguments: kill-ns <namespace>"

kubectl proxy &>/dev/null &
PROXY_PID=$!
killproxy () {
	kill $PROXY_PID
}
trap killproxy EXIT

sleep 1 # give the proxy a second

kubectl get namespace "$PROJECT" -o json | jq 'del(.spec.finalizers[] | select("kubernetes"))' | curl -s -k -H "Content-Type: application/json" -X PUT -o /dev/null --data-binary @- http://localhost:8001/api/v1/namespaces/$PROJECT/finalize && echo "Killed namespace: $PROJECT"

# proxy will get killed by the trap ```
`#!/bin/bash

###############################################################################
# Copyright (c) 2018 Red Hat Inc
#
# See the NOTICE file(s) distributed with this work for additional
# information regarding copyright ownership.
#
# This program and the accompanying materials are made available under the
# terms of the Eclipse Public License 2.0 which is available at
# http://www.eclipse.org/legal/epl-2.0
#
# SPDX-License-Identifier: EPL-2.0
###############################################################################

set -eo pipefail

die() { echo "$*" 1>&2 ; exit 1; }

need() {
	which "$1" &>/dev/null || die "Binary '$1' is missing but required"
}

# checking pre-reqs

need "jq"
need "curl"
need "kubectl"

PROJECT="$1"
shift

test -n "$PROJECT" || die "Missing arguments: kill-ns <namespace>"

kubectl proxy &>/dev/null &
PROXY_PID=$!
killproxy () {
	kill $PROXY_PID
}
trap killproxy EXIT

sleep 1 # give the proxy a second

kubectl get namespace "$PROJECT" -o json | jq 'del(.spec.finalizers[] | select("kubernetes"))' | curl -s -k -H "Content-Type: application/json" -X PUT -o /dev/null --data-binary @- http://localhost:8001/api/v1/namespaces/$PROJECT/finalize && echo "Killed namespace: $PROJECT"

# proxy will get killed by the trap`


usage :  ./kill-kube-ns <name of the namespace>

**Comment by mikerain** [comment]
> > same issue on 1.15 , nothing worked in my case.
> > ```
> > kubectl delete ns test
> > ```
> 
> > Error from server (Conflict): Operation cannot be fulfilled on namespaces "test": The system is ensuring all content is removed from this namespace.  Upon completion, this namespace will automatically be purged by the system.
> 
> have you tried to delete it using the following script ?
> 
> ```
> #!/bin/bash
> 
> ###############################################################################
> # Copyright (c) 2018 Red Hat Inc
> #
> # See the NOTICE file(s) distributed with this work for additional
> # information regarding copyright ownership.
> #
> # This program and the accompanying materials are made available under the
> # terms of the Eclipse Public License 2.0 which is available at
> # http://www.eclipse.org/legal/epl-2.0
> #
> # SPDX-License-Identifier: EPL-2.0
> ###############################################################################
> 
> set -eo pipefail
> 
> die() { echo "$*" 1>&2 ; exit 1; }
> 
> need() {
> 	which "$1" &>/dev/null || die "Binary '$1' is missing but required"
> }
> 
> # checking pre-reqs
> 
> need "jq"
> need "curl"
> need "kubectl"
> 
> PROJECT="$1"
> shift
> 
> test -n "$PROJECT" || die "Missing arguments: kill-ns <namespace>"
> 
> kubectl proxy &>/dev/null &
> PROXY_PID=$!
> killproxy () {
> 	kill $PROXY_PID
> }
> trap killproxy EXIT
> 
> sleep 1 # give the proxy a second
> 
> kubectl get namespace "$PROJECT" -o json | jq 'del(.spec.finalizers[] | select("kubernetes"))' | curl -s -k -H "Content-Type: application/json" -X PUT -o /dev/null --data-binary @- http://localhost:8001/api/v1/namespaces/$PROJECT/finalize && echo "Killed namespace: $PROJECT"
> 
> # proxy will get killed by the trap ```
> `#!/bin/bash
> 
> ###############################################################################
> # Copyright (c) 2018 Red Hat Inc
> #
> # See the NOTICE file(s) distributed with this work for additional
> # information regarding copyright ownership.
> #
> # This program and the accompanying materials are made available under the
> # terms of the Eclipse Public License 2.0 which is available at
> # http://www.eclipse.org/legal/epl-2.0
> #
> # SPDX-License-Identifier: EPL-2.0
> ###############################################################################
> 
> set -eo pipefail
> 
> die() { echo "$*" 1>&2 ; exit 1; }
> 
> need() {
> 	which "$1" &>/dev/null || die "Binary '$1' is missing but required"
> }
> 
> # checking pre-reqs
> 
> need "jq"
> need "curl"
> need "kubectl"
> 
> PROJECT="$1"
> shift
> 
> test -n "$PROJECT" || die "Missing arguments: kill-ns <namespace>"
> 
> kubectl proxy &>/dev/null &
> PROXY_PID=$!
> killproxy () {
> 	kill $PROXY_PID
> }
> trap killproxy EXIT
> 
> sleep 1 # give the proxy a second
> 
> kubectl get namespace "$PROJECT" -o json | jq 'del(.spec.finalizers[] | select("kubernetes"))' | curl -s -k -H "Content-Type: application/json" -X PUT -o /dev/null --data-binary @- http://localhost:8001/api/v1/namespaces/$PROJECT/finalize && echo "Killed namespace: $PROJECT"
> 
> # proxy will get killed by the trap`
> ```
> 
> ```
> usage :  ./kill-kube-ns <name of the namespace>
> ```

it works for me.

**Comment by gxm651182644** [comment]
> in my case, it was cert-managaer ns, which stuck in termination state. It could not be delete because of CRDs which was created before but was not seen by cmd ` kubectl get all` but link to the script which has been provided by @difabion do the job. thx @difabion

Have you solved it?

**Comment by sghose0007** [comment]
Running - kubectl replace --raw "/api/v1/namespaces/demo-app-02/finalize" -f ./demo-app-02.json getting the below error
**Error from server (Forbidden): namespaces "demo-app-02" is forbidden: User "sso:Administrator@vsphere.local" cannot update resource "namespaces/finalize" in API group "" in the namespace "demo-app-02"**

Running this - curl -k -H "Content-Type: application/json" -X PUT --data-binary @demo-app-02.json https://127.0.0.1:8001/api/v1/namespaces/demo-app-02/finalize

Getting the below error 
**curl: (35) SSL received a record that exceeded the maximum permissible length.**

Any solution to remove the namespace..

**Comment by joekendal** [comment]
None of these suggestions worked and after realising I could be looking at wasting hours trying to fix it, I came to the conclusion that if many equally capable devs had tried before me then I would be wasting my time for something so insignificant. Just deleting the kubernetes data and settings worked fine.

**Comment by tp6m4fu6250071** [comment]
> > same issue on 1.15 , nothing worked in my case.
> > ```
> > kubectl delete ns test
> > ```
> 
> > Error from server (Conflict): Operation cannot be fulfilled on namespaces "test": The system is ensuring all content is removed from this namespace.  Upon completion, this namespace will automatically be purged by the system.
> 
> have you tried to delete it using the following script ?
> 
> ```
> #!/bin/bash
> 
> ###############################################################################
> # Copyright (c) 2018 Red Hat Inc
> #
> # See the NOTICE file(s) distributed with this work for additional
> # information regarding copyright ownership.
> #
> # This program and the accompanying materials are made available under the
> # terms of the Eclipse Public License 2.0 which is available at
> # http://www.eclipse.org/legal/epl-2.0
> #
> # SPDX-License-Identifier: EPL-2.0
> ###############################################################################
> 
> set -eo pipefail
> 
> die() { echo "$*" 1>&2 ; exit 1; }
> 
> need() {
> 	which "$1" &>/dev/null || die "Binary '$1' is missing but required"
> }
> 
> # checking pre-reqs
> 
> need "jq"
> need "curl"
> need "kubectl"
> 
> PROJECT="$1"
> shift
> 
> test -n "$PROJECT" || die "Missing arguments: kill-ns <namespace>"
> 
> kubectl proxy &>/dev/null &
> PROXY_PID=$!
> killproxy () {
> 	kill $PROXY_PID
> }
> trap killproxy EXIT
> 
> sleep 1 # give the proxy a second
> 
> kubectl get namespace "$PROJECT" -o json | jq 'del(.spec.finalizers[] | select("kubernetes"))' | curl -s -k -H "Content-Type: application/json" -X PUT -o /dev/null --data-binary @- http://localhost:8001/api/v1/namespaces/$PROJECT/finalize && echo "Killed namespace: $PROJECT"
> 
> # proxy will get killed by the trap ```
> `#!/bin/bash
> 
> ###############################################################################
> # Copyright (c) 2018 Red Hat Inc
> #
> # See the NOTICE file(s) distributed with this work for additional
> # information regarding copyright ownership.
> #
> # This program and the accompanying materials are made available under the
> # terms of the Eclipse Public License 2.0 which is available at
> # http://www.eclipse.org/legal/epl-2.0
> #
> # SPDX-License-Identifier: EPL-2.0
> ###############################################################################
> 
> set -eo pipefail
> 
> die() { echo "$*" 1>&2 ; exit 1; }
> 
> need() {
> 	which "$1" &>/dev/null || die "Binary '$1' is missing but required"
> }
> 
> # checking pre-reqs
> 
> need "jq"
> need "curl"
> need "kubectl"
> 
> PROJECT="$1"
> shift
> 
> test -n "$PROJECT" || die "Missing arguments: kill-ns <namespace>"
> 
> kubectl proxy &>/dev/null &
> PROXY_PID=$!
> killproxy () {
> 	kill $PROXY_PID
> }
> trap killproxy EXIT
> 
> sleep 1 # give the proxy a second
> 
> kubectl get namespace "$PROJECT" -o json | jq 'del(.spec.finalizers[] | select("kubernetes"))' | curl -s -k -H "Content-Type: application/json" -X PUT -o /dev/null --data-binary @- http://localhost:8001/api/v1/namespaces/$PROJECT/finalize && echo "Killed namespace: $PROJECT"
> 
> # proxy will get killed by the trap`
> ```
> 
> ```
> usage :  ./kill-kube-ns <name of the namespace>
> ```

got the issue on EKS (**Kubernetes version: 1.15**) 
kubectl version:
```
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.4", GitCommit:"c96aede7b5205121079932896c4ad89bb93260af", GitTreeState:"clean", BuildDate:"2020-06-17T11:41:22Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"15+", GitVersion:"v1.15.11-eks-065dce", GitCommit:"065dcecfcd2a91bd68a17ee0b5e895088430bd05", GitTreeState:"clean", BuildDate:"2020-07-16T01:44:47Z", GoVersion:"go1.12.17", Compiler:"gc", Platform:"linux/amd64"}
```

The kill-kube-ns script works for me. Thank you.

**Comment by tangyouze** [comment]
I use ``` kubectl edit namespace ****``` and remove the finalizer part, and save.

https://github.com/rancher/rancher/issues/14715#issuecomment-407900343

**Comment by sharma-raj** [comment]
I am unable to delete ns, still showing in terminating state, there's no filed were mentioned the finalizer option. can some please help to resolve the problem?

cat fleet-system.json
{
    "apiVersion": "v1",
    "kind": "Namespace",
    "metadata": {
        "annotations": {
            "cattle.io/status": "{\"Conditions\":[{\"Type\":\"ResourceQuotaInit\",\"Status\":\"True\",\"Message\":\"\",\"LastUpdateTime\":\"2021-01-08T09:14:26Z\"},{\"Type\":\"InitialRolesPopulated\",\"Status\":\"True\",\"Message\":\"\",\"LastUpdateTime\":\"2021-01-08T09:14:31Z\"}]}",
            "field.cattle.io/projectId": "c-s8hs9:p-5bk2c",
            "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"v1\",\"kind\":\"Namespace\",\"metadata\":{\"annotations\":{},\"name\":\"fleet-system\"},\"spec\":{\"finalizers\":[]}}\n",
            "lifecycle.cattle.io/create.namespace-auth": "true",
            "objectset.rio.cattle.io/applied": "{\"apiVersion\":\"v1\",\"kind\":\"Namespace\",\"metadata\":{\"annotations\":{\"objectset.rio.cattle.io/id\":\"fleet-agent-bootstrap\"},\"labels\":{\"objectset.rio.cattle.io/hash\":\"f399d0b310fbfb28e9667312fdc7a33954e2b8c8\"},\"name\":\"fleet-system\"}}",
            "objectset.rio.cattle.io/id": "fleet-agent-bootstrap"
        },
        "creationTimestamp": "2021-01-08T09:14:25Z",
        "deletionGracePeriodSeconds": 0,
        "deletionTimestamp": "2021-01-11T08:41:43Z",
        "finalizers": [
            "controller.cattle.io/namespace-auth"
        ],
        "labels": {
            "field.cattle.io/projectId": "p-5bk2c",
            "objectset.rio.cattle.io/hash": "f399d0b310fbfb28e9667312fdc7a33954e2b8c8"
        },
        "managedFields": [
            {
                "apiVersion": "v1",
                "fieldsType": "FieldsV1",
                "fieldsV1": {
                    "f:metadata": {
                        "f:labels": {
                            "f:field.cattle.io/projectId": {}
                        }
                    }
                },
                "manager": "agent",
                "operation": "Update",
                "time": "2021-01-08T09:14:25Z"
            },
            {
                "apiVersion": "v1",
                "fieldsType": "FieldsV1",
                "fieldsV1": {
                    "f:metadata": {
                        "f:annotations": {
                            ".": {},
                            "f:objectset.rio.cattle.io/applied": {},
                            "f:objectset.rio.cattle.io/id": {}
                        },
                        "f:labels": {
                            ".": {},
                            "f:objectset.rio.cattle.io/hash": {}
                        }
                    }
                },
                "manager": "fleetcontroller",
                "operation": "Update",
                "time": "2021-01-08T09:14:25Z"
            },
            {
                "apiVersion": "v1",
                "fieldsType": "FieldsV1",
                "fieldsV1": {
                    "f:metadata": {
                        "f:annotations": {
                            "f:cattle.io/status": {},
                            "f:field.cattle.io/projectId": {},
                            "f:lifecycle.cattle.io/create.namespace-auth": {}
                        },
                        "f:finalizers": {
                            ".": {},
                            "v:\"controller.cattle.io/namespace-auth\"": {}
                        }
                    }
                },
                "manager": "rancher",
                "operation": "Update",
                "time": "2021-01-08T09:14:30Z"
            },
            {
                "apiVersion": "v1",
                "fieldsType": "FieldsV1",
                "fieldsV1": {
                    "f:status": {
                        "f:phase": {}
                    }
                },
                "manager": "kube-controller-manager",
                "operation": "Update",
                "time": "2021-01-11T08:42:10Z"
            },
            {
                "apiVersion": "v1",
                "fieldsType": "FieldsV1",
                "fieldsV1": {
                    "f:metadata": {
                        "f:annotations": {
                            "f:kubectl.kubernetes.io/last-applied-configuration": {}
                        }
                    }
                },
                "manager": "kubectl",
                "operation": "Update",
                "time": "2021-01-11T10:56:28Z"
            }
        ],
        "name": "fleet-system",
        "resourceVersion": "88004811",
        "selfLink": "/api/v1/namespaces/fleet-system",
        "uid": "af190fda-969c-4fbc-b233-81fa52449411"
    },
    "spec": {},
    "status": {
        "conditions": [
            {
                "lastTransitionTime": "2021-01-11T08:41:48Z",
                "message": "All resources successfully discovered",
                "reason": "ResourcesDiscovered",
                "status": "False",
                "type": "NamespaceDeletionDiscoveryFailure"
            },
            {
                "lastTransitionTime": "2021-01-11T08:41:48Z",
                "message": "All legacy kube types successfully parsed",
                "reason": "ParsedGroupVersions",
                "status": "False",
                "type": "NamespaceDeletionGroupVersionParsingFailure"
            },
            {
                "lastTransitionTime": "2021-01-11T08:41:48Z",
                "message": "All content successfully deleted, may be waiting on finalization",
                "reason": "ContentDeleted",
                "status": "False",
                "type": "NamespaceDeletionContentFailure"
            },
            {
                "lastTransitionTime": "2021-01-11T08:42:09Z",
                "message": "All content successfully removed",
                "reason": "ContentRemoved",
                "status": "False",
                "type": "NamespaceContentRemaining"
            },
            {
                "lastTransitionTime": "2021-01-11T08:41:48Z",
                "message": "All content-preserving finalizers finished",
                "reason": "ContentHasNoFinalizers",
                "status": "False",
                "type": "NamespaceFinalizersRemaining"
            }
        ],
        "phase": "Terminating"
    }
}

**Comment by KornKalle** [comment]
I came across this issue when cleaning up our staging cluster, which our developers use a lot.
We use a cluster with nodes provisioned through rancher at Digitalocean.
For the other people ending here after googling this issue and looking for an easy way to remove these namespaces, i will leave the shell script I've written for these cases here, please use it with care:

```#! /bin/bash
# This script deletes namespaces created through rancher with dangling finalizers
namespace=undefined

# path to your kube config file, e.g.: ~/.kube/config
kubeconfig=
# URL of your rancher, e.g.: https://rancher.example.com
rancher_url=
# ID of the cluster, will be found in the URL of the cluster start page: https://rancher.example.com/c/<CLUSTER_ID>/monitoring
cluster_id=

# Your Rancher Bearer Token generated at 'APIs & Keys' in Rancher
RANCHER_BEARER=

# Ask which namespace will be delete
echo "Enter Namespace you want to delete:"
read namespace

echo "Get Namespace $namespace"
kubectl --kubeconfig $kubeconfig get ns $namespace -o json > $namespace.json

# Removes the whole "Spec" block of the namespace
echo "Removing spec block"
sed -i -e '/\"spec\"/,/}/ d; /^$/d' $namespace.json

# Push namespace back, will be deleted immediately if already dangling
echo "Send edited json file back to rancher"
curl -k -H "Content-Type: application/json" -H "Authorization: Bearer $RANCHER_BEARER" -X PUT --data-binary @$namespace.json $rancher_url/k8s/clusters/$cluster_id/api/v1/namespaces/$namespace/finalize```
